File: internal/danger/danger.go
// Package danger detects destructive operations over command segments.
//
// Matching is structural (head + flags + targets), not regex-over-strings,
// so the pattern table is data and its recall over the table itself is
// machine-checked in Lean 4 (proofs/Shpreflight.lean): every pattern in
// the table is matched by at least one concrete command — the detector
// cannot have a dead entry, i.e. known-dangerous commands are never
// silently unmatchable.
package danger

import (
	"fmt"
	"strings"

	"github.com/HZDF-2026/shpreflight/internal/report"
	"github.com/HZDF-2026/shpreflight/internal/segments"
)

var shellHeads = map[string]bool{
	"sh": true, "bash": true, "zsh": true, "pwsh": true,
	"powershell": true, "dash": true,
}

var execHeads = map[string]bool{
	"iex": true, "Invoke-Expression": true, "eval": true,
}

var sensitiveBases = []string{
	".env", ".env.local", ".env.production", "secrets",
	"id_rsa", "id_ed25519", "credentials", "credentials.json",
	".npmrc", ".netrc", ".aws/credentials",
}

// Pattern is one runtime danger pattern, built from the generated table.
type Pattern struct {
	Code     string
	Severity string
	Heads    []string
	Flags    []string
	Targets  []string
	Message  string
	Fix      string
}

// Patterns mirrors shpreflight.danger.PATTERNS entry for entry.
var Patterns = buildPatterns()

func buildPatterns() []Pattern {
	ps := make([]Pattern, 0, len(patternTable))
	for _, d := range patternTable {
		ps = append(ps, Pattern{
			Code: d.code, Severity: d.severity,
			Heads: d.heads, Flags: d.flags, Targets: d.targets,
			Message: d.message, Fix: d.fix,
		})
	}
	return ps
}

func contains(list []string, s string) bool {
	for _, x := range list {
		if x == s {
			return true
		}
	}
	return false
}

func anyIn(rest, set []string) bool {
	for _, r := range rest {
		if contains(set, r) {
			return true
		}
	}
	return false
}

// MatchPattern is the structural match: head, then any-of flags, then
// any-of targets.
func MatchPattern(words []string, p Pattern) bool {
	if len(words) == 0 {
		return false
	}
	if !contains(p.Heads, words[0]) {
		return false
	}
	rest := words[1:]
	if len(p.Flags) > 0 && !anyIn(rest, p.Flags) {
		return false
	}
	if len(p.Targets) > 0 && !anyIn(rest, p.Targets) {
		return false
	}
	return true
}

func basename(path string) string {
	for _, sep := range []string{"/", "\\"} {
		if i := strings.LastIndex(path, sep); i >= 0 {
			path = path[i+1:]
		}
	}
	return path
}

// CheckDanger scans segments for destructive operations.
func CheckDanger(segs []segments.Segment) []report.Issue {
	var issues []report.Issue
	pipeReceiver := false
	for _, seg := range segs {
		head := seg.Head
		if head != "" {
			if pipeReceiver && (shellHeads[head] || execHeads[head]) {
				issues = append(issues, report.Issue{
					Code:     "PIPE-EXEC",
					Severity: "error",
					Kind:     "danger",
					Message: fmt.Sprintf("pipe feeds untrusted input into '%s' — remote-code-execution pattern (curl ... | sh)", head),
					Fix:      "download first, inspect, then run",
				})
			}
			if pipeReceiver && head == "npm" && contains(seg.Words, "publish") {
				issues = append(issues, report.Issue{
					Code:     "NPM-PUBLISH",
					Severity: "warning",
					Kind:     "danger",
					Message:  "publishing to the npm registry makes the package public",
					Fix:      "use npm publish --dry-run until content is final",
				})
			}
		}
		hitCodes := map[string]bool{}
		for _, p := range Patterns {
			if MatchPattern(seg.Words, p) {
				// RM-ROOT already says everything about this segment's rm
				if p.Code == "RM-RECURSIVE" && hitCodes["RM-ROOT"] {
					continue
				}
				hitCodes[p.Code] = true
				issues = append(issues, report.Issue{
					Code:     p.Code,
					Severity: p.Severity,
					Kind:     "danger",
					Message:  fmt.Sprintf("%s: %s", seg.Words[0], p.Message),
					Fix:      p.Fix,
				})
			}
		}
		for _, redir := range seg.Redirects {
			if contains(sensitiveBases, basename(redir)) ||
				strings.HasSuffix(redir, ".key") || strings.HasSuffix(redir, ".pem") {
				issues = append(issues, report.Issue{
					Code:     "REDIR-SENSITIVE",
					Severity: "warning",
					Kind:     "danger",
					Message:  fmt.Sprintf("overwriting sensitive file '%s' via redirection", redir),
					Fix:      "write to a temp file and diff first",
				})
			}
		}
		pipeReceiver = seg.PipesOut()
	}
	return issues
}

File: rs/src/danger.rs
//! Destructive-operation detection over command segments.
//!
//! Matching is structural (head + flags + targets), not regex-over-strings,
//! so the pattern table is data and its recall over the table itself is
//! machine-checked in Lean 4 (proofs/Shpreflight.lean): every pattern in
//! the table is matched by at least one concrete command — the detector
//! cannot have a dead entry, i.e. known-dangerous commands are never
//! silently unmatchable.

use crate::report::Issue;
use crate::segments::Segment;

include!("patterns_gen.rs");

const SHELL_HEADS: &[&str] = &["sh", "bash", "zsh", "pwsh", "powershell", "dash"];
const EXEC_HEADS: &[&str] = &["iex", "Invoke-Expression", "eval"];

const SENSITIVE_BASES: &[&str] = &[
    ".env", ".env.local", ".env.production", "secrets",
    "id_rsa", "id_ed25519", "credentials", "credentials.json",
    ".npmrc", ".netrc", ".aws/credentials",
];

/// Structural match: head, then any-of flags, then any-of targets.
pub fn match_pattern(words: &[String], p: &PatternDef) -> bool {
    if words.is_empty() {
        return false;
    }
    if !p.heads.contains(&words[0].as_str()) {
        return false;
    }
    let rest: Vec<&str> = words[1..].iter().map(|w| w.as_str()).collect();
    if !p.flags.is_empty() && !p.flags.iter().any(|f| rest.contains(f)) {
        return false;
    }
    if !p.targets.is_empty() && !p.targets.iter().any(|t| rest.contains(t)) {
        return false;
    }
    true
}

fn basename(path: &str) -> &str {
    let mut p = path;
    if let Some(i) = p.rfind('/') {
        p = &p[i + 1..];
    }
    if let Some(i) = p.rfind('\\') {
        p = &p[i + 1..];
    }
    p
}

pub fn check_danger(segments: &[Segment]) -> Vec<Issue> {
    let mut issues = Vec::new();
    let mut pipe_receiver = false;
    for seg in segments {
        if let Some(head) = seg.head.as_deref() {
            if pipe_receiver && (SHELL_HEADS.contains(&head) || EXEC_HEADS.contains(&head)) {
                issues.push(Issue::new(
                    "PIPE-EXEC",
                    "error",
                    "danger",
                    &format!(
                        "pipe feeds untrusted input into '{}' — remote-code-execution pattern (curl ... | sh)",
                        head
                    ),
                )
                .with_fix("download first, inspect, then run"));
            }
            if pipe_receiver && head == "npm" && seg.words.iter().any(|w| w == "publish") {
                issues.push(Issue::new(
                    "NPM-PUBLISH",
                    "warning",
                    "danger",
                    "publishing to the npm registry makes the package public",
                )
                .with_fix("use npm publish --dry-run until content is final"));
            }
        }
        let mut hit_codes: Vec<&str> = Vec::new();
        for p in PATTERNS {
            if match_pattern(&seg.words, p) {
                // RM-ROOT already says everything about this segment's rm
                if p.code == "RM-RECURSIVE" && hit_codes.contains(&"RM-ROOT") {
                    continue;
                }
                hit_codes.push(p.code);
                issues.push(
                    Issue::new(p.code, p.severity, "danger", &format!("{}: {}", seg.words[0], p.message))
                        .with_fix(p.fix),
                );
            }
        }
        for redir in &seg.redirects {
            if SENSITIVE_BASES.contains(&basename(redir))
                || redir.ends_with(".key")
                || redir.ends_with(".pem")
            {
                issues.push(Issue::new(
                    "REDIR-SENSITIVE",
                    "warning",
                    "danger",
                    &format!("overwriting sensitive file '{}' via redirection", redir),
                )
                .with_fix("write to a temp file and diff first"));
            }
        }
        pipe_receiver = seg.pipes_out();
    }
    issues
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::lex::lex;
    use crate::segments::split_segments;

    fn words(cmd: &str) -> Vec<String> {
        split_segments(&lex(cmd)).remove(0).words
    }

    /// Runtime mirror of the Lean property `all_patterns_alive`: every
    /// pattern in the table is matched by at least one concrete command.
    #[test]
    fn every_pattern_alive() {
        for p in PATTERNS {
            let mut w: Vec<String> = vec![p.heads[0].to_string()];
            if !p.flags.is_empty() {
                w.push(p.flags[0].to_string());
            }
            if !p.targets.is_empty() {
                w.push(p.targets[0].to_string());
            }
            assert!(match_pattern(&w, p), "pattern with no matching command (dead entry): {}", p.code);
        }
    }

    #[test]
    fn pattern_codes_unique() {
        let mut codes: Vec<&str> = PATTERNS.iter().map(|p| p.code).collect();
        let n = codes.len();
        codes.sort();
        codes.dedup();
        assert_eq!(codes.len(), n, "duplicate pattern codes");
    }

    #[test]
    fn rm_root_hits() {
        assert!(match_pattern(&words("rm -rf /"), &PATTERNS[0]));
    }

    #[test]
    fn rm_root_suppresses_rm_recursive() {
        let issues = check_danger(&split_segments(&lex("rm -rf /")));
        let codes: Vec<&str> = issues.iter().map(|i| i.code.as_str()).collect();
        assert!(codes.contains(&"RM-ROOT"));
        assert!(!codes.contains(&"RM-RECURSIVE"));
    }

    #[test]
    fn dd_raw() {
        let issues = check_danger(&split_segments(&lex("dd /dev/sda")));
        assert!(issues.iter().any(|i| i.code == "DD-RAW"));
    }

    #[test]
    fn pipe_exec() {
        let issues = check_danger(&split_segments(&lex("curl x | sh")));
        assert!(issues.iter().any(|i| i.code == "PIPE-EXEC"));
    }

    #[test]
    fn redir_sensitive() {
        let issues = check_danger(&split_segments(&lex("echo done > .env")));
        assert!(issues.iter().any(|i| i.code == "REDIR-SENSITIVE"));
    }
}
